express.1
.TH EXPRESS "1" "March 2013" "express 3.1.0" "User Commands"
.SH NAME
express \- Sinatra inspired web development framework
.SH SYNOPSIS
Usage: express [options]
.SH DESCRIPTION
This Node.js module is a fast, unopinionated, minimalist web development
framework inspired by Sinatra.

The Express philosophy is to provide small, robust tooling for HTTP servers.
Making it a great solution for single page applications, web sites, hybrids,
or public HTTP APIs.

Built on Connect you can use only what you need, and nothing more,
applications can be as big or as small as you like, even a single file.
Express does not force you to use any specific ORM or template engine. With
support for over 14 template engines via Consolidate.js you can quickly craft
your perfect framework.
.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
output usage information
.TP
\fB\-V\fR, \fB\-\-version\fR
output the version number
.TP
\fB\-s\fR, \fB\-\-sessions\fR
add session support
.TP
\fB\-e\fR, \fB\-\-ejs\fR
add ejs engine support (defaults to jade)
.TP
\fB\-J\fR, \fB\-\-jshtml\fR
add jshtml engine support (defaults to jade)
.TP
\fB\-H\fR, \fB\-\-hogan\fR
add hogan.js engine support
.TP
\fB\-c\fR, \fB\-\-css\fR <engine>
add stylesheet <engine> support (less|stylus) (defaults to plain css)
.TP
\fB\-f\fR, \fB\-\-force\fR
force on non\-empty directory
.SH AUTHORS
The main author of expresso is TJ Holowaychuk <\fBtj@vision-media.ca\fR>.
For a more complete list of contributors, see the GitHub page at
<\fBhttps://github.com/visionmedia/express\fR>
.SH REPORTING BUGS
Please report bugs on the GitHub issue tracker:
<\fBhttps://github.com/visionmedia/express/issues\fR>


History.md
unreleased
==========

  * deps: debug@2.6.3
    - Fix: `DEBUG_MAX_ARRAY_LENGTH`
  * deps: finalhandler@~1.0.1
    - Fix missing `</html>` in HTML document
    - deps: debug@2.6.3
  * deps: proxy-addr@~1.1.4
    - deps: ipaddr.js@1.3.0
  * deps: send@0.15.2
    - deps: debug@2.6.4
    - deps: ms@1.0.0
  * deps: serve-static@1.12.2
    - deps: send@0.15.2
  * deps: type-is@~1.6.15
    - deps: mime-types@~2.1.15
  * deps: vary@~1.1.1
    - perf: hoist regular expression

4.15.2 / 2017-03-06
===================

  * deps: qs@6.4.0
    - Fix regression parsing keys starting with `[`

4.15.1 / 2017-03-05
===================

  * deps: send@0.15.1
    - Fix issue when `Date.parse` does not return `NaN` on invalid date
    - Fix strict violation in broken environments
  * deps: serve-static@1.12.1
    - Fix issue when `Date.parse` does not return `NaN` on invalid date
    - deps: send@0.15.1

4.15.0 / 2017-03-01
===================

  * Add debug message when loading view engine
  * Add `next("router")` to exit from router
  * Fix case where `router.use` skipped requests routes did not
  * Remove usage of `res._headers` private field
    - Improves compatibility with Node.js 8 nightly
  * Skip routing when `req.url` is not set
  * Use `%o` in path debug to tell types apart
  * Use `Object.create` to setup request & response prototypes
  * Use `setprototypeof` module to replace `__proto__` setting
  * Use `statuses` instead of `http` module for status messages
  * deps: debug@2.6.1
    - Allow colors in workers
    - Deprecated `DEBUG_FD` environment variable set to `3` or higher
    - Fix error when running under React Native
    - Use same color for same namespace
    - deps: ms@0.7.2
  * deps: etag@~1.8.0
    - Use SHA1 instead of MD5 for ETag hashing
    - Works with FIPS 140-2 OpenSSL configuration
  * deps: finalhandler@~1.0.0
    - Fix exception when `err` cannot be converted to a string
    - Fully URL-encode the pathname in the 404
    - Only include the pathname in the 404 message
    - Send complete HTML document
    - Set `Content-Security-Policy: default-src 'self'` header
    - deps: debug@2.6.1
  * deps: fresh@0.5.0
    - Fix false detection of `no-cache` request directive
    - Fix incorrect result when `If-None-Match` has both `*` and ETags
    - Fix weak `ETag` matching to match spec
    - perf: delay reading header values until needed
    - perf: enable strict mode
    - perf: hoist regular expressions
    - perf: remove duplicate conditional
    - perf: remove unnecessary boolean coercions
    - perf: skip checking modified time if ETag check failed
    - perf: skip parsing `If-None-Match` when no `ETag` header
    - perf: use `Date.parse` instead of `new Date`
  * deps: qs@6.3.1
    - Fix array parsing from skipping empty values
    - Fix compacting nested arrays
  * deps: send@0.15.0
    - Fix false detection of `no-cache` request directive
    - Fix incorrect result when `If-None-Match` has both `*` and ETags
    - Fix weak `ETag` matching to match spec
    - Remove usage of `res._headers` private field
    - Support `If-Match` and `If-Unmodified-Since` headers
    - Use `res.getHeaderNames()` when available
    - Use `res.headersSent` when available
    - deps: debug@2.6.1
    - deps: etag@~1.8.0
    - deps: fresh@0.5.0
    - deps: http-errors@~1.6.1
  * deps: serve-static@1.12.0
    - Fix false detection of `no-cache` request directive
    - Fix incorrect result when `If-None-Match` has both `*` and ETags
    - Fix weak `ETag` matching to match spec
    - Remove usage of `res._headers` private field
    - Send complete HTML document in redirect response
    - Set default CSP header in redirect response
    - Support `If-Match` and `If-Unmodified-Since` headers
    - Use `res.getHeaderNames()` when available
    - Use `res.headersSent` when available
    - deps: send@0.15.0
  * perf: add fast match path for `*` route
  * perf: improve `req.ips` performance

4.14.1 / 2017-01-28
===================

  * deps: content-disposition@0.5.2
  * deps: finalhandler@0.5.1
    - Fix exception when `err.headers` is not an object
    - deps: statuses@~1.3.1
    - perf: hoist regular expressions
    - perf: remove duplicate validation path
  * deps: proxy-addr@~1.1.3
    - deps: ipaddr.js@1.2.0
  * deps: send@0.14.2
    - deps: http-errors@~1.5.1
    - deps: ms@0.7.2
    - deps: statuses@~1.3.1
  * deps: serve-static@~1.11.2
    - deps: send@0.14.2
  * deps: type-is@~1.6.14
    - deps: mime-types@~2.1.13

4.14.0 / 2016-06-16
===================

  * Add `acceptRanges` option to `res.sendFile`/`res.sendfile`
  * Add `cacheControl` option to `res.sendFile`/`res.sendfile`
  * Add `options` argument to `req.range`
    - Includes the `combine` option
  * Encode URL in `res.location`/`res.redirect` if not already encoded
  * Fix some redirect handling in `res.sendFile`/`res.sendfile`
  * Fix Windows absolute path check using forward slashes
  * Improve error with invalid arguments to `req.get()`
  * Improve performance for `res.json`/`res.jsonp` in most cases
  * Improve `Range` header handling in `res.sendFile`/`res.sendfile`
  * deps: accepts@~1.3.3
    - Fix including type extensions in parameters in `Accept` parsing
    - Fix parsing `Accept` parameters with quoted equals
    - Fix parsing `Accept` parameters with quoted semicolons
    - Many performance improvments
    - deps: mime-types@~2.1.11
    - deps: negotiator@0.6.1
  * deps: content-type@~1.0.2
    - perf: enable strict mode
  * deps: cookie@0.3.1
    - Add `sameSite` option
    - Fix cookie `Max-Age` to never be a floating point number
    - Improve error message when `encode` is not a function
    - Improve error message when `expires` is not a `Date`
    - Throw better error for invalid argument to parse
    - Throw on invalid values provided to `serialize`
    - perf: enable strict mode
    - perf: hoist regular expression
    - perf: use for loop in parse
    - perf: use string concatination for serialization

lib/express.core.js

// Express - Core - Copyright TJ Holowaychuk <tj@vision-media.ca> (MIT Licensed)

process.mixin(GLOBAL, require("sys"))
var http = require("http")

Express = { 
  version: '0.0.1',
  utilities: {},
  modules: [],
  routes: [],
  responseCodes : {
    100: 'Continue',
    101: 'Switching Protocols',
    200: 'OK',
    201: 'Created',
    202: 'Accepted',
    203: 'Non-Authoritative Information',
    204: 'No Content',
    205: 'Reset Content',
    206: 'Partial Content',
    300: 'Multiple Choices',
    301: 'Moved Permanently',
    302: 'Moved Temporarily',
    303: 'See Other',
    304: 'Not Modified',
    305: 'Use Proxy',
    400: 'Bad Request',
    401: 'Unauthorized',
    402: 'Payment Required',
    403: 'Forbidden',
    404: 'Not Found',
    405: 'Method Not Allowed',
    406: 'Not Acceptable',
    407: 'Proxy Authentication Required',
    408: 'Request Time-out',
    409: 'Conflict',
    410: 'Gone',
    411: 'Length Required',
    412: 'Precondition Failed',
    413: 'Request Entity Too Large',
    414: 'Request-URI Too Large',
    415: 'Unsupported Media Type',
    500: 'Internal Server Error',
    501: 'Not Implemented',
    502: 'Bad Gateway',
    503: 'Service Unavailable',
    504: 'Gateway Time-out',
    505: 'HTTP Version not supported'
  },
  
  // --- Response
  
  response : {
    headers : {},
    cookie : {}
  },
  
  // --- Settings
  
  settings : {
    basepath : '/',
    defaultRoute : {
      callback : function() {
        Express.halt('Page or file cannot be found', 'Not Found')
      }
    }
  },
  
  // --- Modules
  
  RedirectHelpers : {
    init : function() {
      Express.home = Express.settings.basepath
    },
    
    onRequest : {
      'set back to referrer' : function() {
        Express.back = 
          Express.requestHeader('Referer') ||
            Express.requestHeader('Referrer')
      }
    }
  },
  
  ContentLength : {
    onResponse : {
      'set content length' : function() {
        Express.header('Content-Length', (Express.response.body || '').length)
      }
    }
  },
  
  DefaultContentType : {
    onRequest : {
      'set response content type to text/html' : function() {
        Express.header('Content-Type', 'text/html')
      }
    }
  },
  
  BodyDecoder : {
    onRequest : {
      'parse urlencoded bodies' : function() {
        switch (Express.requestHeader('Content-Type')) {
          case 'application/x-www-form-urlencoded':
            Express.request.uri.params = Express.parseParams(decodeURIComponent(Express.request.body))
            break
          
          case 'application/json':
            Express.request.uri.params = Express.jsonDecode(Express.request.body)
            break
        }
      }
    }
  },
  
  MethodOverride : {
    onRequest : {
      'set HTTP method to _method when present' : function() {
        if (method = Express.param('_method'))
          Express.request.method = method.toUpperCase()
      }
    }
  },
  
  Logger : {
    onResponse : {
      'output log data' : function(){
        puts('"' + Express.request.method + ' ' + Express.request.uri.path +
             ' HTTP/' + Express.request.httpVersion + '" - ' + Express.response.status)
      }
    }
  },
  
  /**
   * Invoke hook _name_ with _args_, returning
   * an array of return values.
   *
   * @param  {string} name
   * @param  {...} args
   * @return {array}
   * @api public
   */
  
  hook : function(name, args) {
    var results = []
    for (var i = 0, len = this.modules.length; i < len; ++i)
      for (var key in this.modules[i][name])
        results.push(this.modules[i][name][key].apply(this.modules[i], this.toArray(arguments, 1))) 
    return results
  },
  
  /**
   * Invoke hook _name_ with _immutable_ and _args_, returning
   * the _immutable_ variable once modified by each hook implementation.
   * When implementing these hooks, the function must return a value.
   *
   * @param  {string} name
   * @param  {mixed} immutable
   * @param  {...} args
   * @return {mixed}
   * @api public
   */
  
  hookImmutable : function(name, immutable, args) {
    for (var i = 0, len = this.modules.length; i < len; ++i)
      for (var key in this.modules[i][name])
        immutable = this.modules[i][name][key].call(this.modules[i], immutable)
    return immutable
  },
  
  /**
   * Add _module_.
   *
   * @param  {hash} module
   * @param  {...} args
   * @api public
   */
  
  addModule : function(module, args) {
    Express.settings[module.name] = {}

    if ('init' in module) 
      module.init.apply(module, Express.toArray(arguments, 1))

    if ('settings' in module) {
      if (!module.name) throw 'module name is required when using .settings'
      for (var name in module.settings)
        Express.settings[module.name][name] = module.settings[name]
    }
          
    if ('utilities' in module)
      for (var name in module.utilities)
        Express.utilities[name] = module.utilities[name]
    
    Express.modules.push(module)
  },
  
  /**
   * Start Express; binding to _host_ / _port_.
   *
   * @param {int} port
   * @param {string} host
   * @api public
   */
  
  start : function(port, host) {
    this.server.listen(port || 3000, host, this.server.callback)
  },
  
  server : {
    
    /**
     * Listen to _host_ / _port_ with _callback_.
     *
     * @param {int} port
     * @param {string} host
     * @param {function} callback
     * @api private
     */
    
    listen : function(port, host, callback) {
      http.createServer(callback).listen(port, host)
      puts('Express started at http://' + (host || '127.0.0.1') + ':' + port + '/')
    },
    
    /**
     * Request callback.
     *
     * @param  {object} request
     * @param  {object} response
     * @api private
     */
    
    callback : function(request, response){
      request.body = request.body || ''
      request.setBodyEncoding('utf8')
      request.addListener('body', function(chunk) { request.body += chunk })
      request.addListener('complete', function(){
        Express.home = Express.settings.basepath
        Express.response.body = null
        Express.response.status = 200
        Express.request = request
        Express.originalResponse = response
        request.uri.params = Express.parseNestedParams(request.uri.params)
        request.uri.path = Express.normalizePath(request.uri.path)
        Express.hook('onRequest')
        try { Express.response.body = Express.callRouteFor(request) }
        catch (e) { Express.fail(e) }
        if (typeof Express.response.body == 'string')
          Express.server.finished()
      })
    },
  
    finished: function(body) {
      if (body) Express.response.body = body
      Express.hook('onResponse')
      Express.originalResponse.sendHeader(Express.response.status, Express.response.headers)
      Express.originalResponse.sendBody(Express.response.body || '')
      Express.originalResponse.finish()
    }
  },
  
  /**
   * Fail Express with exception _e_.
   *
   * @param  {object} e
   * @api private
   */

  fail : function(e) {
    this.hook('onFailure', e)
    this.header('Content-Type', 'text/plain')
    this.response.body = e.message ? e.name + ': ' + e.message : e.toString()
    this.response.status = 500
  },
  
  /**
   * JSON encode _object_
   *
   * @param  {mixed} object
   * @return {string}
   * @api public
   */
  
  jsonEncode : function(object) {
    return JSON.stringify(object)
  },
  
  /**
   * JSON decode _string_.
   *
   * @param  {string} string
   * @return {mixed}
   * @api public
   */
   
   jsonDecode : function(string) {
     return JSON.parse(string)
   },
   
  /**
   * Redirect to _uri_.
   *
   * @param  {string} uri
   * @api public
   */
  
  redirect : function(uri) {
    this.header('Location', uri)
    this.halt("Page or file has moved to `" + uri + "'", 'Moved Temporarily')
  },
  
  /**
   * Set response status to _value_. Where value may be
   * a case-insensitive string such as 'Not Found', 'forbidden',
   * or a numeric HTTP response status code.
   *
   * @param  {int, string} value
   * @return {int}
   * @api public
   */
  
  status : function(value) {
    if (value == undefined) return
    if (typeof value == 'number') return this.response.status = value
    for (var code in this.responseCodes)
      if (this.responseCodes[code].toLowerCase() == value.toLowerCase())
        return this.response.status = parseInt(code)
  },

  /**
   * Return path segment _n_.
   *
   * @param  {int} n
   * @return {string}
   * @api public
   */
  
  arg : function(n) {
    return this.request.uri.path.split('/')[n]
  },
  
  /**
   * Set response _status_ and _body_. Status
   * defaults to 200.
   *
   * @param  {string} body
   * @param  {string, int} status
   * @api public
   */
  
  halt : function(body, status) {
    this.status(status || 200)
    Express.server.finished(this.response.body = body)
  },
  
  /**
   * Get or set response _name_ header.
   *
   * @param  {string} name
   * @param  {string} value
   * @return {string}
   * @api public
   */
  
  header : function(name, value) {
    return value ? 
      this.response.headers[name.toLowerCase()] = value :
        this.response.headers[name.toLowerCase()]
  },
  
  /**
   * Get request header _name_ optionally providing _default_
   * when _name_ is not present.
   *
   * @param  {string} name
   * @param  {string} default
   * @return {string}
   * @api public
   */
  
  requestHeader : function(name, defaultValue) {
    return this.request.headers[name.toLowerCase()] || defaultValue
  },
      
  /**
   * Return param _key_'s value or _defaultValue_ when
   * the param has not been set.
   *
   * @param  {string} key
   * @param  {mixed} defaultValue
   * @return {mixed}
   * @api public
   */
  
  param : function(key, defaultValue) {
    return this.request.uri.params[key] || defaultValue
  },
  
  /**
   * Parse params _string_ into a nested hash.
   *
   * @param  {string} string
   * @return {hash}
   * @api public
   */
  
  parseParams : function(string) {
    var params = {}, pairs = string.split('&'), pair
    for (var i = 0, len = pairs.length; i < len; ++i) {
      pair = pairs[i].split('=')
      params[pair[0]] = pair[1]
    }
    return this.parseNestedParams(params)
  },
  
  /**
   * Parse _params_, splitting { user[info][name] : 'tj' } into
   * { user : { info : { name : 'tj' }}}
   *
   * @param  {hash} params
   * @return {hash}
   * @api public
   */
  
  parseNestedParams : function(params) {
    for (var key in params)
      if (parts = key.split('['))
        if (parts.length > 1)
          for (var i = 0, prop = params, len = parts.length; i < len; ++i) {
            var name = parts[i].replace(']', '')
            if (i == len - 1) {
              prop[name] = params[key]
              prop = params, delete params[key]
            }
            else {
              prop = prop[name] = prop[name] || {}
            }
          }
              
    return params
  },
  
  /**
   * Parse _cookie_ string into an object (hash).
   *
   * @param  {string} cookie
   * @return {hash}
   * @api public
   */
  
  parseCookie : function(cookie) {
    var hash = {}
    if (!cookie) return hash
    var attrs = cookie.split(/\s*;\s*/)
    for (var i = 0; i < attrs.length; ++i)
      hash[attrs[i].split('=')[0]] = decodeURIComponent(attrs[i].split('=')[1])
    return hash
  },
  
  /**
   * Normalize a path. Strips trailing / leading whitespace
   * and slashes ('/').
   *
   * @param  {string} path
   * @return {string}
   * @api public
   */
  
  normalizePath : function(path) {
    if (typeof path != 'string') return path
    return path.replace(/^(\s|\/)*|(\s|\/)*$/g, '')
  },
  
  /**
   * Normalize _path_. When a RegExp it is simply returned,
   * otherwise a string is converted to a regular expression
   * surrounded by ^$. So /user/:id/edit would become:
   *
   * \^/user/(.*?)\/edit\$/
   *
   * Each param key (:id) will be captured and placed in the
   * params array, so param('id') would give the string captured.
   *
   * @param  {string} path
   * @return {regexp}
   * @api public
   */
  
  pathToRegexp : function(path) {
    Express.regexpKeys = []
    if (path.constructor == RegExp) return path
    path = path.replace(/:(\w+)/g, function(_, key){
      Express.regexpKeys.push(key)
      return '(.*?)'
    })
    return new RegExp('^' + this.escapeRegexp(path, '/ [ ]') + '$', 'i')
  },
  
  /**
   * Escape RegExp _chars_ in _string_. Where _chars_ 
   * defaults to regular expression special characters.
   *
   * _chars_ should be a space delimited list of characters,
   * for example '[ ] ( )'.
   *
   * @param  {string} string
   * @param  {string} chars
   * @return {Type}
   * @api public
   */
  
  escapeRegexp : function(string, chars) {
    var specials = (chars || '/ . * + ? | ( ) [ ] { } \\').split(' ').join('|\\')
    return string.replace(new RegExp('(\\' + specials + ')', 'g'), '\\$1')
  },
  
  /**
   * Return the contents of a function body.
   *
   * @param  {function} body
   * @return {string}
   * @api public
   */
  
  contentsOf : function(body) {
    if (typeof body != 'function') throw "Express.contentsOf(): `" + body + "' is not a function"
    return body.toString().match(/^[^\{]*{((.*\n*)*)}/m)[1]
  },
  
  /**
   * Eval _string_ in context of Express.
   *
   * @param  {string} string
   * @return {mixed}
   * @api public
   */
  
  eval : function(string) {
    return eval('with (Express){ with (Express.utilities){ ' + string + '}}')
  },
      
  /**
   * Escape special characters in _html_.
   *
   * @param  {string} html
   * @return {string}
   * @api public
   */
  
  escape : function(html) {
    return html.
      replace(/&/g, '&amp;').
      replace(/"/g, '&quot;').
      replace(/</g, '&lt;').
      replace(/>/g, '&gt;')
  },
  
  /**
   * Convert native array-like objects into an
   * array with optional _offset_.
   *
   * @param  {object} arr
   * @param  {int} offset
   * @return {array}
   * @api public
   */
  
  toArray : function(arr, offset) {
    return Array.prototype.slice.call(arr, offset)
  },
  
  /**
   * Check if _route_ matches provides the proper
   * encoding types listed in _request_'s Accept header.
   *
   * @param  {hash} route
   * @param  {hash} request
   * @return {bool}
   * @api private
   */
  
  routeProvides : function(route, request) {
    if (!route.options.provides) return true
    if (route.options.provides.constructor != Array) 
      route.options.provides = [route.options.provides]
    for (var i = 0, len = route.options.provides.length; i < len; ++i)
      if (request.headers.accept && request.headers.accept.match(route.options.provides[i]))
        return true
    return false
  },
  
  /**
   * Attept to match and call a route for 
   * the given _request_, returning the data
   * returned by the route callback.
   *
   * @param  {object} request
   * @return {mixed}
   * @api private
   */
  
  callRouteFor : function(request) {
    var route = this.findRouteFor(request) || this.settings.defaultRoute
    if (route.keys)
      for (var i = 0, len = route.keys.length; i < len; ++i) 
        this.request.uri.params[route.keys[i]] = Express.captures[i+1] || null
    var body = this.eval(Express.contentsOf(route.callback))
    if (typeof body != 'string') body = null
    return body
  },
  
  /**
   * Attempt to find and return a route matching _request_.
   *
   * @param  {object} request
   * @return {object}
   * @api private
   */
  
  findRouteFor : function(request) {
    for (var i = 0, len = this.routes.length; i < len; ++i)
      if (this.routeMatches(this.routes[i], request))
        return this.routes[i]
  },
  
  /**
   * Check if _route_ matches _request_
   *
   * @param  {object} route
   * @param  {object} request
   * @return {bool}
   * @api private
   */
  
  routeMatches : function(route, request) {
    if (request.method.toLowerCase() != route.method) return false
    switch (route.path.constructor) {
      case String: return route.path == request.uri.path
      case RegExp: return !! (Express.captures = request.uri.path.match(route.path))
    }
  },
  
  /**
   * Return a routing function for _method_.
   *
   * @param  {string} method
   * @return {function}
   * @api private
   */
  
  routeFunctionFor : function(method) {
    return function(path, options, callback) {
      if (options.constructor == Function) callback = options, options = {}
      path = Express.pathToRegexp(Express.normalizePath(path))
      var route = {
        keys : Express.regexpKeys,
        path : path,
        method : method,
        options : options,
        callback : callback
      }
      Express.routes.push(route)
      Express.hook('onRouteAdded', route)
    }
  }
}

// --- Routing API

get = Express.routeFunctionFor('get')
post = Express.routeFunctionFor('post')
del = Express.routeFunctionFor('delete')
put = Express.routeFunctionFor('put')
use = Express.addModule

// --- Core Modules

use(Express.BodyDecoder)
use(Express.MethodOverride)
use(Express.ContentLength)
use(Express.DefaultContentType)
use(Express.RedirectHelpers)
